src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp

```diff
@@ -241,6 +224,264 @@ void C2_MacroAssembler::fast_unlock(Register objectReg, Register boxReg, Registe
+
+    // Check if recursive.
+    subw(t, top, oopSize);
+    ldr(t, Address(rthread, t));
+    cmp(obj, t);
+    br(Assembler::EQ, unlocked);
+
+    // Not recursive.
+    // Load Mark.
+    ldr(mark, Address(obj, oopDesc::mark_offset_in_bytes()));
+
+    // Check header for monitor (0b10).
+    tbnz(mark, exact_log2(markWord::monitor_value), inflated);
+
+    // Try to unlock. Transition lock bits 0b00 => 0b01
+    assert(oopDesc::mark_offset_in_bytes() == 0, "required to avoid lea");
+    orr(t, mark, markWord::unlocked_value);
+    // Release to satisfy the JMM.
+    cmpxchg(/*addr*/ obj, /*expected*/ mark, /*new*/ t, Assembler::xword,
+            /*acquire*/ false, /*release*/ true, /*weak*/ false, noreg);
+    br(Assembler::EQ, unlocked);
+
+    // Load link store conditional exclusive failed.
+    // Restore lock-stack and handle the unlock in runtime.
+    DEBUG_ONLY(str(obj, Address(rthread, top));)
+    addw(top, top, oopSize);
+    str(top, Address(rthread, JavaThread::lock_stack_top_offset()));
+    b(slow_path);
+  }
+
+
+  { // Handle inflated monitor.
+    bind(inflated_load_monitor);
+    ldr(mark, Address(obj, oopDesc::mark_offset_in_bytes()));
+#ifdef ASSERT
+    tbnz(mark, exact_log2(markWord::monitor_value), inflated);
+    stop("Fast Unlock not monitor");
+#endif
+
+    bind(inflated);
+
+#ifdef ASSERT
+    Label check_done;
+    subw(top, top, oopSize);
+    cmpw(top, in_bytes(JavaThread::lock_stack_base_offset()));
+    br(Assembler::LT, check_done);
+    ldr(t, Address(rthread, top));
+    cmp(obj, t);
+    br(Assembler::NE, inflated);
+    stop("Fast Unlock lock on stack");
+    bind(check_done);
+#endif
+
+    // mark contains the tagged ObjectMonitor*.
+    const Register monitor = mark;
+    const uintptr_t monitor_tag = markWord::monitor_value;
+
+    // Untag the monitor.
+    sub(monitor, mark, monitor_tag);
+
+    const Register recursions = t2;
+    Label not_recursive;
+
+    // Check if recursive.
+    ldr(recursions, Address(monitor, ObjectMonitor::recursions_offset()));
+    cbz(recursions, not_recursive);
+
+    // Recursive unlock.
+    sub(recursions, recursions, 1u);
+    str(recursions, Address(monitor, ObjectMonitor::recursions_offset()));
+    // Set flag == EQ
+    cmp(recursions, recursions);
+    b(unlocked);
+
+    bind(not_recursive);
+
+    Label release;
+    const Register owner_addr = t2;
+
+    // Compute owner address.
+    lea(owner_addr, Address(monitor, ObjectMonitor::owner_offset()));
+
+    // Check if the entry lists are empty.
+    ldr(rscratch1, Address(monitor, ObjectMonitor::EntryList_offset()));
+    ldr(t, Address(monitor, ObjectMonitor::cxq_offset()));
+    orr(rscratch1, rscratch1, t);
+    cmp(rscratch1, zr);
+    br(Assembler::EQ, release);
+
+    // The owner may be anonymous and we removed the last obj entry in
+    // the lock-stack. This loses the information about the owner.
+    // Write the thread to the owner field so the runtime knows the owner.
```

Comment:
Is this necessary here also?  Previous checks and slow path code in the runtime has already set the owner, if I understand correctly.

Reply:
After popping the last oop of the lock stack we do the `tbnz(mark, exact_log2(markWord::monitor_value), inflated);` check. If this happen the owner will be anonymous. 

Other solutions would be either:
  1. Push the oop back and jump to the runtime. (Would make C2 anonymous owner agnostic).
  2. Fix the owner only in this control flow, not in every inflated slow path exit.

The first seems alright as well. It is more like what x86 evolved into doing (where it elides this specific check). 
Both solutions make the inflated unlock cleaner removes a branch (can branch directly to the slow path).
The second does seems does create a more complex entry to the inflated unlock, does not seem worth it.

Reply:
I didn't see that this isn't like the x86 code that pushes the object back on the lock stack.  This seems fine and the comment is helpful.